File: U16VirtualComPort/ringBuffer.f
#define GBUF_SIZE 128

void checksize( ringBuffer_t *pr, uint8_t *label );

uint8_t gbuf[GBUF_SIZE];
ringBuffer_t ring;
ringBuffer_t *pr;

void checksize( ringBuffer_t *pr, uint8_t *label ) {
	printf ( "LABEL:%s ", label );
	if ( isRingBufferEmpty(pr) == TRUE ) {
		printf ( "Ring Buffer is empty\n" );
	} else if ( isRingBufferFull(pr) == TRUE ) {
		printf ( "Ring Buffer is full=%03d\n", getRingBufferDataSize(pr) );
	} else {
		printf ( "Ring Buffer has data=%03d\n", getRingBufferDataSize(pr) );
	}
}

void main( void ) {
	uint8_t data,i;
	uint8_t datas[20];
	pr = &ring;

	initRingBuffer( pr, gbuf, GBUF_SIZE );
	checksize ( pr, "1" );
	setRingBuffer( pr, 'a' );
	checksize ( pr, "2" );
	getRingBuffer( pr, &data);
	checksize ( pr, "3" );

	setRingBuffers( pr, "0123456789" , 10 );
	checksize ( pr, "4" );
	getRingBuffers( pr, datas, 10 );
	datas[10] = '\0';
	printf ( "%s\n", datas );
	checksize ( pr, "5" );

	for ( i = 0 ; i < 13 ; i++ ) {
		setRingBuffers( pr, "0123456789" , 10 );
		checksize ( pr, "6+" );
	}
	for ( i = 0 ; i < 12 ; i++ ) {
		getRingBuffers( pr, datas, 10 );
		datas[10] = '\0';
		printf ( "%s\n", datas );
		checksize ( pr, "6-" );
	}
	checksize ( pr, "7" );
	for ( i = 0 ; i < sizeof( datas ); i++ ) { datas[i] = '\0'; }
	getRingBuffers( pr, datas, 10 );
	printf ( "%s\n", datas );
}

File: U16KeyboardAndMouse/ringBuffer.c
#include "stdio.h"
#include "typedef.h"
#include "ringBuffer.h"

void initRingBuffer( ringBuffer_t *ring, uint8_t *buffer, uint16_t size ){
	ring->buf = buffer;
	ring->rp = 0;
	ring->wp = 0;
	ring->size = size;
}

bool_t setRingBuffer( ringBuffer_t *ring, uint8_t data )
{
	bool_t ret = TRUE;
	
	// disable interrupt
	if ( ((ring->wp+1) % ring->size) != ring->rp ){
		ring->buf[ring->wp] = data;
		ring->wp = (ring->wp + 1) % ring->size;
		ret = 1;
	}
	// enable interrupt
	return ret;
}

bool_t getRingBuffer( ringBuffer_t *ring, uint8_t *data )
{
	bool_t ret = FALSE;
	
	// disable interrupt
	if ( (ring->rp != ring->wp ) ){
		*data = ring->buf[ring->rp];
		ring->rp = (ring->rp + 1) % ring->size;
		ret = TRUE;
	}
	// enable interrupt

	return ret;
}

uint16_t setRingBuffers( ringBuffer_t *ring, uint8_t *data, uint16_t size )
{
	uint16_t i;
	uint8_t *pt;

	disableInterrupt();
	pt = data;
	
	for ( i = 0 ; i < size ; i++ ) {
		if ( setRingBuffer( ring, *pt ) == FALSE ) {
			break;
		}
		pt++;
	}
	enableInterrupt();
	return i;
}

uint16_t getRingBuffers( ringBuffer_t *ring, uint8_t *data, uint16_t size )
{
	uint16_t i;
	uint8_t *pt;

	disableInterrupt();
	pt = data;
	
	for ( i = 0 ; i < size ; i++ ) {
		if ( getRingBuffer( ring, pt ) == 0 ) {
			break;
		}
		pt++;
	}
	enableInterrupt();
	return i;
}

uint16_t getRingBufferDataSize( ringBuffer_t *ring ) {
	uint16_t size;
	if ( ring->wp < ring->rp ) {
		size = ( ring->wp + ring->size ) - ring->rp;
	} else {
		size = ring->wp - ring->rp;
	}
	return size;
}

bool_t isRingBufferEmpty( ringBuffer_t *ring )
{
	bool_t ret = FALSE;
	if ( ring->wp == ring->rp ) {
		ret = TRUE;
	}
	return ret;
}

bool_t isRingBufferFull( ringBuffer_t *ring )
{
	bool_t ret = FALSE;
	if ( ((ring->wp+1) % ring->size) == ring->rp ) {
		ret = TRUE;
	}
	return ret;
}

#ifdef FEATURE_TESTCODE
#include "ringBuffer.f"
#endif /* FEATURE_TESTCODE */

